components/brave_ads/content/browser/search_result_ad/search_result_ad_redirect_throttle.cc

```diff
@@ -0,0 +1,112 @@
+  DCHECK(web_contents);
+
+  SessionID tab_id = sessions::SessionTabHelper::IdForTab(web_contents);
+  content::RenderFrameHost* opener_rfh = web_contents->GetOpener();
+  if (!opener_rfh) {
+    opener_rfh = web_contents->GetOriginalOpener();
+  }
+  if (opener_rfh) {
+    content::WebContents* original_web_contents =
+        content::WebContents::FromRenderFrameHost(opener_rfh);
+    tab_id = sessions::SessionTabHelper::IdForTab(original_web_contents);
+  }
+
+  return tab_id;
+}
+
+}  // namespace
+
+// static
+std::unique_ptr<SearchResultAdRedirectThrottle>
+SearchResultAdRedirectThrottle::MaybeCreateThrottleFor(
+    SearchResultAdService* search_result_ad_service,
+    const network::ResourceRequest& request,
+    content::WebContents* web_contents) {
+  DCHECK(web_contents);
+
+  if (!search_result_ad_service || !request.request_initiator ||
+      !request.has_user_gesture || !request.is_outermost_main_frame ||
+      request.resource_type !=
+          static_cast<int>(blink::mojom::ResourceType::kMainFrame)) {
+    return nullptr;
+  }
+
+  if (!brave_search::IsAllowedHost(request.request_initiator->GetURL()) ||
+      !IsSearchResultAdClickedConfirmationUrl(request.url)) {
+    return nullptr;
+  }
+
+  SessionID tab_id = GetTabId(web_contents);
+  if (!tab_id.is_valid()) {
+    return nullptr;
+  }
+
+  return std::make_unique<SearchResultAdRedirectThrottle>(
+      search_result_ad_service, tab_id);
+}
+
+SearchResultAdRedirectThrottle::SearchResultAdRedirectThrottle(
+    SearchResultAdService* search_result_ad_service,
+    SessionID tab_id)
+    : search_result_ad_service_(search_result_ad_service), tab_id_(tab_id) {
+  DCHECK(search_result_ad_service_);
+  DCHECK(tab_id_.is_valid());
+}
+
+SearchResultAdRedirectThrottle::~SearchResultAdRedirectThrottle() = default;
+
+void SearchResultAdRedirectThrottle::WillStartRequest(
+    network::ResourceRequest* request,
+    bool* defer) {
+  DCHECK(request);
+  DCHECK(request->has_user_gesture);
+  DCHECK(request->is_outermost_main_frame);
+  DCHECK_EQ(request->resource_type,
+            static_cast<int>(blink::mojom::ResourceType::kMainFrame));
+  DCHECK(request->request_initiator);
+  DCHECK(brave_search::IsAllowedHost(request->request_initiator->GetURL()));
+
+  const std::string creative_instance_id =
+      GetClickedSearchResultAdCreativeInstanceId(*request);
+  if (creative_instance_id.empty()) {
+    return;
+  }
+
+  const absl::optional<GURL> search_result_ad_target_url =
+      search_result_ad_service_->MaybeTriggerSearchResultAdClickedEvent(
+          creative_instance_id, tab_id_);
+  if (!search_result_ad_target_url) {
+    return;
+  }
+  DCHECK(search_result_ad_target_url->is_valid());
+  DCHECK(search_result_ad_target_url->SchemeIs(url::kHttpsScheme));
+
+  request->url = *search_result_ad_target_url;
+  request->credentials_mode = network::mojom::CredentialsMode::kOmit;
+  request->referrer_policy = net::ReferrerPolicy::NO_REFERRER;
```

Comment:
Added `credentials_mode` and `referrer_policy` modification as suggested here: https://github.com/brave/security/issues/876#issuecomment-1139044261

cc @ShivanKaul @diracdeltas @bridiver

Reply:
@aseren not sure that is necessary. Just to verify, this url is the ad landing page, right? The actual confirmation happens in the background somewhere from MaybeTriggerSearchResultAdClickedEvent, correct?

Reply:
actually looking more closely I think this functionality is not correct. We should not be navigating to the click confirmation link here, we should be navigating directly to the landing page and the confirmation should happen in the background from a non-profile context

Reply:
so after talking to @pes10k we don't need to do anything special here other than insert the token, but there are other issues here because we have all this extra infrastructure and complications that aren't necessary if we're just following the click-through url. We don't need to parse ad info from the page and associate it with the tab, we can just put everything we need in the url.

Reply:
> we should be navigating directly to the landing page and the confirmation should happen in the background from a non-profile context

just a reminder if this functionality is enabled in Tor windows, then any background requests associated with a confirmation that happens in a Tor window needs to still go through Tor; otherwise a network observer can link the user's real IP address to the ID of an ad that they saw in a Tor window

Reply:
@diracdeltas don't worry, it will crash in Tor https://github.com/brave/brave-core/pull/13444/files#r892925877 ;)

Reply:
seriously though we're sticking with the redirect at least for now

Reply:
also seriously please add a test for Tor and private

Reply:
actually we do have tests for private tab: https://github.com/brave/brave-core/pull/13444/files#diff-e2eb5fe1f2ab08f0a20e5e835b489e48211466cdf25bac21263315e165ffc508R432

and SearchResultAdService won't be created for incognito, see factory code:
https://github.com/brave/brave-core/blob/93a3167ac3722808b044d06d46b7ab727382aa9d/browser/brave_ads/search_result_ad/search_result_ad_service_factory.cc

Reply:
The way to prevent it from being created in incognito is to return null in SearchResultAdServiceFactory::BuildServiceInstanceFor